kronmult_gpu/kronmult.cu

```diff
@@ -0,0 +1,207 @@
+#include <device_launch_parameters.h>
+#include <type_traits>
+#include "kronmult.cuh"
+
+/*
+ * computes number^power for integers
+ * does not care about performances
+ * does not use std::pow as it does an implicit float conversion that could lead to rounding errors for large numbers
+ */
+__host__ int pow_int(const int number, const int power)
+{
+    if(power == 0) return 1;
+    return number * pow_int(number, power-1);
+}
+
+/*
+ * converts row and col indices into a single index for a matrix stored in col-major
+ * `stride` is usually the number of rows of the matrix
+ */
+__device__ __forceinline__ constexpr int colmajor(const int row, const int col, const int stride)
+{
+    return row + col*stride;
+}
+
+/*
+ * computes output = input^T
+ *
+ * `input` is a `matrix_size` by `matrix_size` square matrix of stride `input_stride`
+ * `output` is a `matrix_size` by `matrix_size` square matrix of stride `matrix_size`
+ *
+ * WARNING: the matrices are assumed to be stored in col-major order
+ */
+template<typename T>
+__device__ void transpose(const T input[], T output[], const int matrix_size, const int input_stride)
+{
+    for(int i = threadIdx.x; i < matrix_size*matrix_size; i+=blockDim.x)
+    {
+        const int c = i / matrix_size;
+        const int r = i - c*matrix_size;
+        output[colmajor(r, c, matrix_size)] = input[colmajor(c, r, input_stride)];
+    }
+}
+
+/*
+ * Computes Y = X^T * M^T
+ *
+ * X is a `size_M` by `nb_col_X` matrix of stride `size_M`
+ * M_transposed is a `size_M` by `size_M` matrix of stride `size_M` that contains a precomputed M^T
+ * Y is a `nb_col_X` by `size_M` matrix of stride `nb_col_X`
+ *
+ * WARNING: the matrices are assumed to be stored in col-major order
+ */
+template<typename T>
+__device__ void multiply_transpose(const T X[], const int nb_col_X,
+                                   const T M_transposed[], const int size_M,
+                                   T Y[])
+{
+    // strided loop, each thread threadIdx.x manages the inputs i such that threadIdx.x % t==0
+    for(int i = threadIdx.x; i < nb_col_X*size_M; i+=blockDim.x)
+    {
+        // extracts the column and row number for the current thread
+        const int colX = i / size_M;
+        const int rowM = i - colX*size_M;
+        // computes the dot product to fill the [colX,rowM] cell of the matrix
+        T dotprod = 0.;
+        for(int k=0; k < size_M; k++)
+        {
+            dotprod += X[colmajor(k,colX,size_M)] * M_transposed[colmajor(k,rowM,size_M)];
+        }
+        // this sync is there to synchronise the threads for significantly improved performance in float
+        // it does not impact correctness
+        if(std::is_same<float, T>::value) __syncthreads();
```

Comment:
Why add the `constexpr`? (the condition is already evaluated at compile time) Additional clarity?

Reply:
Adding it triggers a `this functionality requiers C++17` warning and no nvcc flag seems to be able to remove it.
I think I will leave it as is (no `constexpr`), unless I find a flag to diseable the warning, since it is elided at compile time without the `constexpr`.

Reply:
@efdazedo @dlg0 ASGarD already requires the C++17 standard. May we set the same minimum for kronmult?

Reply:
The C++ is compiled in C++17, CUDA is the only part impacted.

Reply:
This is now fixed.